lib/vrl/stdlib/src/parse_host.rs

```diff
@@ -0,0 +1,164 @@
+use std::collections::BTreeMap;
+use tldextract::*;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseHost;
+
+impl Function for ParseHost {
+    fn identifier(&self) -> &'static str {
+        "parse_host"
+    }
+
+    fn summary(&self) -> &'static str {
+        "parse a host string"
+    }
+
+    fn usage(&self) -> &'static str {
+        indoc! {r#"
+            Parses the provided `value` into TLD, regiestered domain, and subdomains.
+
+            By default included private domains.
+        "#}
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "private_domains",
+                kind: kind::BOOLEAN,
+                required: false,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "parse host",
+                source: r#"parse_host!("www.example.com")"#,
+                result: Ok(indoc! {r#"
+                    {
+                        "domain": "example",
+                        "subdomain": "www",
+                        "suffix": "com"
+                    }
+                "#}),
+            },
+            Example {
+                title: "ignore private domain",
+                source: r#"parse_host!("internal.us-east-1.elb.amazonaws.com", private_domains: false)"#,
+                result: Ok(indoc! {r#"
+                    {
+                        "domain": "amazonaws",
+                        "subdomain": "internal.us-east-1.elb",
+                        "suffix": "com"
+                    }
+                "#}),
+            },
+        ]
+    }
+
+    fn compile(&self, _state: &state::Compiler, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+
+        let private_domains = arguments
+            .optional("private_domains")
+            .unwrap_or_else(|| expr!(true));
+
+        Ok(Box::new(ParseHostFn {
+            value,
+            private_domains,
+        }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseHostFn {
+    value: Box<dyn Expression>,
+    private_domains: Box<dyn Expression>,
+}
+
+impl Expression for ParseHostFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let string = value.try_bytes_utf8_lossy()?;
+
+        let private_domains = self.private_domains.resolve(ctx)?.try_boolean()?;
+
+        let ext = TldExtractor::new(TldOption {
+            private_domains,
+            ..Default::default()
+        });
```

Comment:
If I'm reading the library's [source](https://github.com/dovahcrow/tldextract-rs/blob/3be7b0edb3bd73779882764a7e6564489798f164/src/cache/local.rs#L14) correctly, then even if we set `local_path: None` (which is the default), it will _still_ use local caching, and read a local file for caching information every time we call `TldExtractor::new`.

Given this, I think we should move the initialization of `TldExtractor` into the `compile` function, to only call it once at boot.

The downside of this is that we can no longer have `private_domains` be dynamically set at runtime. It'll have to be a boolean literal, instead of any expression resolving to a boolean. I think this is in acceptable trade-off.

Reply:
Ah, good point. Would it make sense to try and set that path to something vector already would use as a data dir, or is that going to be too messy for vrl today?

I could imagine possible issues if the default location was read only due to k8s policies or some such.

Reply:
We don't really have any way for VRL to know where it's allowed to store data. I would keep the default for now, until we have a mechanism for the caller to point VRL to some sort of "data dir".

Reply:
An alternative to allow `private_domains` to still be dynamic would be to initialize two clients.